Change `init`'s role in reduce-like functions: remove "neutral element" restriction and guarantee its use

mbauman opened the pull request:
Here's a shot at revamped documentation in light of https://github.com/JuliaLang/julia/issues/49042.  The goal here is ~~threefold~~ fivefold:

* Document that `init` provides the return value if the collections are empty (status quo)
* Document how reductions over empty collections may be errors (status quo)
* Remove the restriction that `init` must be a "neutral element" (**change**)
* Guarantee that `init` is used exactly once (**change** from being completely unspecified)
* Guarantee that `init` is used as the left-most argument in the reduction (**change** from being unspecified)

And in the process, this cleans up the language around associativity a bit.

This is a draft as this language needs to be propagated through to all the "special" reduction functions like `sum` et al.  But it gives us something concrete to consider.

mbauman committed:
Add more clarity to `init`'s role in reduce and mapreduce

mbauman commented on base/reduce.jl:
```diff
@@ -452,28 +461,29 @@ _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f,
-
 Some operations accumulate error. Parallelism will be easier if the reduction can be
-executed in groups. Future versions of Julia might change the algorithm. Note that the
-elements are not reordered if you use an ordered collection.
```
I have no idea what this means?

stevengj replied:
I think the point it is trying to make is that:
1. the associativity of `mapreduce` is not specified — e.g. `mapreduce(+, [a,b,c,d])` might be computed as `a + (b + (c + d))` or `((a + b) + c) + d` or `(a + b) + (c + d)`.     If these are floating-point operations, different associativities can change the roundoff errors.   Future versions of Julia might change the associativity for reasons of performance, accuracy (e.g. [pairwise reductions](https://en.wikipedia.org/wiki/Pairwise_summation) are already used in some cases to improve accuracy), or parallelism (though `mapreduce` is currently serial).
2. for an ordered collection (e.g. an array), although the associativity can change the ordering cannot — Julia does not assume commutativity.   For an unordered collection (e.g. a `Set`), the ordering is also implementation defined (so the results could completely change from version to version if `op` is not commutative).

mbauman replied:
Ohhh I see. I'm so used to associative operators being commutative I didn't even consider that possibility.  But of course.  Any suggestions on how we might make this clearer?  It's something that I could easily see a custom implementation getting wrong — although fortunately the "simple" algorithms one might write probably "happen" to do this correctly.

I had actually used the word "order" above to specify that the sequence of when `f` gets applied to each element in `itr` is not defined.

stevengj replied:
Maybe talk about the "operand ordering".

vtjnash commented:
Also may be related to https://github.com/JuliaLang/julia/pull/44819 and https://github.com/JuliaLang/julia/pull/53461 (for the meaning and role of init in `cumsum` and how type promotion is defined for it)

Seelengrab commented:
> Document that init is used exactly once, and it means that every value in the collection is used as a argument to the reducer op, even for single-argument collections

How does this play together with vectorization & associativity? I vaguely remember that this was a contentious point, since  the result can be very different depending on how often that "initial value" ([seems to be a misnomer](https://github.com/JuliaLang/julia/pull/44819#pullrequestreview-928583796)) is used in reductions. I believe this is the reason the actual number of uses of `init` is currently unspecified.

mbauman commented:
> How does this play together with vectorization & associativity?

That's exactly my motivation here — I want to push forward @stevengj's #52397.  If `init` is specified to only be used exactly once, then only one parallel "chain" can start with `init`.  And that's fine, as long as you don't expect to depend upon `init` causing a cascading promotion.

Looks like https://github.com/JuliaLang/julia/pull/44819 was trying to add similar guarantees to `maximum`.  My intention here is to nail down the language and then copy it to the op-specific reducers like `maximum` and friends.

`accumulate` is slightly more troublesome since (AFAIK) we're missing the analogous associativity-defined verb.

mikmoore commented on base/reduce.jl:
```diff
@@ -281,15 +281,29 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
 
 [`mapreduce`](@ref) is functionally equivalent to calling
-`reduce(op, map(f, itr); init=init)`, but will in general execute faster since no
+`reduce(op, map(f, itrs...)...; init=init)`, but will in general execute faster since no
```
```suggestion
`reduce(op, map(f, itrs...); init=init)`, but will in general execute faster since no
```
This got a little carried away with the dots here.

mikmoore commented:
I'm torn on leaving the reduction position of `init` unspecified. One one hand, specifying it ties our hands on the implementation. On the other, the name `init` is a bit misleading if it's not at the beginning (or at least *a* beginning, assuming we usually opt for pairwise reductions). Although in some implementations it might be easier or more performant to include it at the very end instead (maybe makes initialization easier or cache alignment more likely). In total, I suppose it's easiest to leave explicitly unspecified for now as it can always be made stricter later.

While it can be deduced from the documentation, we could consider making an explicit note warning against using `init` for promotion. Technically, any previous behavior that relied on it was based on an invalid use because a promoting argument is not a neutral element.

mbauman committed:
fix too many dots......

Co-authored-by: mikmoore <95002244+mikmoore@users.noreply.github.com>

mikmoore commented on base/reduce.jl:
```diff
@@ -281,15 +281,29 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+`itr`. Use [`mapfoldl`](@ref) or [`mapfoldr`](@ref) for
+strict left or right associativity and guaranteed invocation of `f` for every value.
+
+If provided, `init` serves as the return value for empty `itrs`. For non-empty iterators,
```
I think we could clarify that `init` is applied directly to the reduction rather than the collection. I.e., we reduce with `init` rather than `f(init)`. But I don't have a clear and concise suggestion for the statement.

mbauman replied:
Yeah, I did initially have a phrase in there about `init` not being transformed by `f` at one point. I agree it'd be nice to include... I just deleted it in the editing process because it was more cumbersome than I liked and I had phrased it in the negative, which always leaves a bad taste in my mouth for docs.

stevengj commented:
We should definitely specify the *ordering* of `init`, i.e. that `init` comes at the "left" (i.e. before the first element) for ordered collections, and hence is effectively used only once.

In principle, we still don't need to specify the associativity, though in most cases a `mapreduce` with `init` will be implemented as `mapfoldl` internally.   But for specific operations like `op === +` we might change the associativity (and be free to add `zero(init)` elsewhere, e.g. at the leaves of pairwise recursion, to ensure the precision of `init` is used for the summation).  But this illustrates the difficulty with not specifying the associativity, since `init` can be used not just to specify the value of empty reductions but also to change the precision.  e.g. what does:
```jl
a = rand(Float16, 1000)
mapreduce(identity, +, a, init=0.0)
```
do?  Is it allowed to compute `0.0 + sum(a)`, i.e. compute the sum in `Float16` precision and then add `0.0` (converting to `Float64`) at the very end?  Or should it be equivalent to `mapfoldl(identity, +, a, init=0.0)`, which performs the sum left-to-right in `Float64` precision?

mbauman commented:
So this is trying to capture the current thought-processes in #49042 — in particular [Jeff's comment](https://github.com/JuliaLang/julia/issues/49042#issuecomment-2021297683).  The case brought up in https://github.com/JuliaLang/julia/pull/44819#issuecomment-1088498161 is helpful in thinking about an associative-but-not-commutative operator: `reduce(*, ["aa","bb","cc"], init="zz")`.

> But this illustrates the difficulty with not specifying the associativity, since `init` can be used not just to specify the value of empty reductions but also to change the precision.

I'm explicitly saying that's a-ok with this PR — and in fact it's exactly what you sign up for with `reduce`.  If you want a cascading promotion like that, well, then you use a function with guaranteed cascading associativity.  Concretely, I don't think your example of `a = rand(Float16, 1000); reduce(+, a, init=0.0)` needs to be computed any differently than `reduce(+, Real[0.0; a])` would be.

stevengj commented:
Okay, but I think we should still document that `init` is used at the left, e.g. `reduce(*, ["aa","bb","cc"], init="zz")` must `"zzaabbcc"`.   As far as I know, `mapreduce` has for a long time assumed `op` is associative, but has never assumed `op` is commutative — it's documented that it can change the associativity but not the ordering.  So an `init` argument doesn't make sense unless you say where it goes.

mikmoore commented:
> (But I'm still worried that re-associating `init` may break code that depended on the current promotion behavior.)

Yes, breaking this would be *annoying* for those that used it successfully. However, it was also explicitly not a valid use (or at least highly dubious) under the previous rules of `?reduce`:

> If provided, the initial value init must be a neutral element for op that will be returned for empty collections. It is unspecified whether init is used for non-empty collections.

We went so far as to specify that `init` could be ignored entirely for non-empty collections, so one "could not" rely upon it being used for promotion or anything else. And optionality aside, those rules require that `init` satisfy a neutrality contract that `op(init, f(x[i]))` "equal" `f(x[i])`. We weren't very specific about exactly what "equal" meant here. `==` seems arbitrary since it can mean so much, but `===` seems too strict because one couldn't reduce collections of mutables. But I could go so far as to imagine that it should satisfy "equal" meaning "`==` and of matching type."

mbauman committed:
Incorporate feedback; define left-most position for init

mbauman commented:
OK, I've updated the documentation changes here with the feedback and incorporated the requirement that `init` be incorporated as the left-most argument to `op`.  BUT: This is now becoming a bigger change in the **specification** of the reduction and whereas my initial intention here was to more match the documentation with the status quo implementations, I'm now thinking that we're probably imposing additional requirements that may not be satisfied by the status quo implementations.

At a minimum CUDA does not follow this specification: https://github.com/JuliaLang/julia/issues/49042#issuecomment-1477821335.  It's unclear to me where to go from here.

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+different methods and between Julia versions. For example, `-` is not
+associative and thus `mapreduce(√, -, [1, 4, 9])` may return either
+`-4.0` or `2.0` depending upon the exact method or version of Julia.
+This is also true of some floating point operations that are typically
+associative, for example `mapreduce(identity, +, [.1, .2, .3])` may return
```
Floating-point operations are typically *not* associative.

Maybe something like:

> Because floating-point roundoff errors typically break associativity, even for operations like `+` that are associative in exact arithmetic, this also means that the roundoff characteristics of `mapreduce` are implementation-defined; for example, …

mbauman replied:
Yeah, I meant that to read that `+` itself is typically associative.  I see how I swung and missed, though :)

mbauman replied:
```suggestion
Because floating-point roundoff errors typically break associativity,
even for operations like + that are associative in exact arithmetic,
this also means that the floating-point errors incurred by mapreduce
are implementation-defined; for example `mapreduce(identity, +, [.1, .2, .3])` may return
```

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+either `0.6` or `0.6000000000000001`.
+
+While the associativity of the reduction is not defined, `mapreduce` does preserve
+the ordering of the iterator for ordered collections.  For example,
```
```suggestion
the ordering of the iterator for ordered collections, so that the result does *not* require `op` to be commutative.  For example,
```

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+the ordering of the iterator for ordered collections.  For example,
+`mapreduce(uppercase, string, ['j','u','l','i','a'])` is guaranteed to always
+return the properly-spelled `"JULIA"` because `Array`s are ordered collections;
+the returned ordering is not guaranteed with an unordered collection like `Set`.
```
```suggestion
in contrast, the operand ordering is not guaranteed with an unordered collection like `Set`.
```

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+
+While the associativity of the reduction is not defined, `mapreduce` does preserve
+the ordering of the iterator for ordered collections.  For example,
+`mapreduce(uppercase, string, ['j','u','l','i','a'])` is guaranteed to always
```
```suggestion
`mapreduce(uppercase, *, ['j','u','l','i','a'])` is guaranteed to always
```

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
 [`map`](@ref).
 
+Some commonly-used operators may have special implementations of a mapped reduction, and
+should be used instead: [`maximum`](@ref)`(itr)`, [`minimum`](@ref)`(itr)`, [`sum`](@ref)`(itr)`,
```
```suggestion
are recommended instead of `mapreduce`: [`maximum`](@ref)`(itr)`, [`minimum`](@ref)`(itr)`, [`sum`](@ref)`(itr)`,
```

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,60 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+julia> mapreduce(identity, +, [.1, .2, .3]) ≈ 0.6
+true
+
+julia> mapreduce(uppercase, string, ['j','u','l','i','a'])
+"JULIA"
+
+julia> mapreduce(uppercase, string, ['j','u','l','i','a'], init="Hello ")
+"Hello JULIA"
```
```suggestion
julia> mapreduce(uppercase, *, ['j','u','l','i','a'])
"JULIA"

julia> mapreduce(uppercase, *, ['j','u','l','i','a'], init="Hello ")
"Hello JULIA"
```

stevengj replied:
(It just seems non-idiomatic to use `string` as the concatenation operator.)

stevengj commented:
(I don't see why parallel reduction requires `init` to be the "neutral" identity element for `op` — if you are subdividing into parallel chunks, just divide into non-empty chunks where `init` isn't needed.)

mbauman commented:
> (I don't see why parallel reduction ala CUDA requires `init` to be the "neutral" identity element for `op` — if you are subdividing into parallel chunks, just divide into non-empty chunks where `init` isn't needed.)

It surely doesn't require it, but it **is** a current implementation that deviates from this PR's proposed documentation.  I don't know how many other implementations are at variance.

stevengj commented:
Fair enough — it seems like the documentation that `init` (formerly `v0`) must be a "neutral" element dates back to Julia 0.4 (#9027), so it's probably not reasonable to change it now, just to clarify what that means.

mbauman commented:
I still think this is a useful exercise here because this is describing what I personally would want `init` to mean and makes concrete what the changes would look like.  There are three points:

> 1. Remove the restriction that init must be a "neutral element"
> 2. Guarantee that init is used exactly once (change from being completely unspecified)
> 3. Guarantee that init is used as the left-most argument in the reduction (change from being unspecified)

Point (1) is tricky, because some common use-cases (included some documented ones) already disregard it.  Points (2) and (3) would be perfectly compatible with the prior documentation if we remove the "guarantee".

So where to go?  We could either decide that this is a "minor change" worth doing given existing usages and implementations (not many packages seem to use a non-neutral init, but I could imagine that this kind of "abuse" would be more common in scripts and non-package code), or we could find a new kwarg name with these tighter semantics, or we could just give up on #49042 for now and clarify the status quo as best we can.

mbauman committed:
Apply suggestions from code review

Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

stevengj commented:
I would also like to remove the counter-intuitive restriction that `init` be a "neutral" element (I've been using Julia for more than a decade and never even realized that this restriction existed).  But I'm not certain that this is a Julia 1.x change.

mikmoore commented:
I think I lost something along the way here. Can someone explain in what sense `mbauman`'s proposed changes are breaking. It seems to me that any use of `reduce` that was valid under the previous restrictions should be valid with these new rules.

(Disregard the rest of my comments if there is a sense is which this is breaking to the *use* of `reduce`.)

Any existing *implementations* of `reduce` might not initially support the new rules. Is that the sense? I suppose it could at least be considered a minor change under that consideration. Public packages are searchable, so CUDA.jl et al may need PRs but that seems doable. Private code that implements custom `reduce` functions is likely to continue to implement the old restrictions (if not a different set entirely, if supporting `init` at all) until updated (probably never). But such private code is unlikely to be used with the new `init` rules, so that seems like a victimless crime.

I don't see this as profoundly different than any other feature expansion (e.g., most anything with a compat notice). Pre-existing implementations may not support the expanded interface.

`init` is nigh-useless under its current specification. It seems that the only thing currently guaranteed is that it is returned for `isempty` collections. A ternary can trivially accomplish without the keyword. It would be unfortunate to indefinitely block making `init` useful because a few packages would need PRs.

mbauman commented:
No, you're exactly right — the "breaking-ness" here is simply that we're changing the rules of the game out from underneath existing implementations.  From the calling side, all of these changes "fit" within the previous documentation.  But it means that previously technically-correct implementations are now immediately "buggy" as soon as the new docs lands.

That said, I do think it's warranted to call this a minor change and still do it, but that doesn't mean it's not without impact.  It'd be good to get some voices from the JuliaFolds folks here (@MasonProtter, @jw3126, @DilumAluthge) — what do y'all think?

mbauman committed:
Add NEWS.md

mbauman commented on NEWS.md:
```diff
@@ -77,6 +77,14 @@ New library features
 
+* The `init` keyword for `reduce` and other reduction functions without guaranteed
+  associativity (`mapreduce`, `maximum`, `minimum`, `sum`, `prod`, `any`, and `all`)
+  now provides greater gaurantees on how its value is incorporated into the reduction:
```
```suggestion
  now provides greater guarantees on how its value is incorporated into the reduction:
```

mbauman committed:
typo fix

jw3126 commented:
I like it. I think some paralell implementations in the ecosystem will still want the neutral element assumption. But IMO it is then reasonable to make it their job to document that.

arhik commented on NEWS.md:
```diff
@@ -79,7 +79,7 @@ Standard library changes
 
 * The `init` keyword for `reduce` and other reduction functions without guaranteed
```
also "keyword" should be "keyword argument" I think. It's pretty much implied when I went through the PR. But First glance at first statement made me think this is introducing new keyword.

MasonProtter commented:
I'm against this direction.

This change would make it so that the `init` keyword argument does not actually "initialize" a reduction, but instead is just the leftmost element of the collection. IMO that's not a particularly useful property compared to an initializer. 

I'd much rather that we make the opposite guarantee: that the `init` kwarg *must* be used for every recursive sub-reduction. e.g. if we have
```
reduce(op, [1,2, 3, 4, 5])
```
which gets split to 
```
op(foldl(op, [1,2,3]), foldl(op, [4, 5]))
```
then the `init` option should be propagated to both `foldl` calls. This can be really useful for things like recovering type stability, or avoiding issues from doing parallel reductions over collections like `Channel` that don't have a predictable size.

mikmoore commented:
I suppose the less breaking thing would be to retain `init`'s current role and neutrality (and extend its guarantee to the beginning of all reduction chains, as suggested). I still think using `init` for promotion is an abuse (and is unreliable under the existing semantics). If one needs promotion, (e.g.) `sum(Float64, x; init=truly_neutral)` is easy to write and works predictably under any conceivable semantics.

If we really wanted, we could introduce a new keyword element to apply at the left (and right?) with the proposed semantics. However, the proposed changes don't do anything that couldn't be implemented simply and more clearly as `isempty(x) ? leftmost : op(leftmost, reduce(op, x))`, so a keyword is probably overkill.

adienes commented:
I'm with @MasonProtter . there are easy enough ways to simply add something to the left of a collection. reducing with an `init` at each subproblem is more difficult (especially if trying to implement a parallel reduction)

maleadt commented:
I haven't really read into the whole problem here, so just responding to the GPU bits.

> (I don't see why parallel reduction ala CUDA requires `init` to be the "neutral" identity element for `op` — if you are subdividing into parallel chunks, just divide into non-empty chunks where `init` isn't needed.)

Just dividing into non-empty chunks is not trivial, because many operations (e.g. tree-style reduction in shared memory, or subgroup intrinsics) are executed cooperatively, i.e., by a number of threads that may not cleanly divide the number of elements. Adding branches to prevent that is generally not a good idea on GPUs.

So right now in GPUArrays.jl we use `init` as a neutral element, even for non-empty inputs. For compatibility with existing code that doesn't specify `init` because the input isn't empty (e.g. `sum(f, ::CuArray)`), we have a `GPUArrays.neutral_element(f, ET)` interface. And as a final kludge, we have our own version of `mapreducedim!` that also takes an `init` argument, because the API-dictated behavior of initializing the output (and then each thread reading back from it to get the neutral element) is too costly.

jonas-schulze commented:
I support the position of @MasonProtter. If a user requires a certain associativity, or that `init` be used at a specific end of the reduction chain, or any other assumption to the order of operations,[^1] that user should rather use the more [imperative](https://en.wikipedia.org/wiki/Imperative_programming) `mapfoldl` or `mapfoldr`. IMO, `mapreduce` must retain its current and [declarative](https://en.wikipedia.org/wiki/Declarative_programming) semantics, in order to support e.g. tree-like reductions[^2] as used by Dagger.[^3]

Maybe the non-associative example could be extended to explicitly allow all of the following to be returned by `reduce(-, [0, 1, 2, 3])`:

```
julia> ((0 - 1) - 2) - 3
-6

julia> 0 - (1 - (2 - 3))
-2

julia> (0 - 1) - (2 - 3)
0
```

[^1]: like a side-effect of the "map" part to be triggered before any reduction is evaluated
[^2]: as mentioned by https://github.com/JuliaLang/julia/pull/53871#discussion_r1541389149
[^3]: see https://github.com/JuliaParallel/Dagger.jl/issues/475 and the [implementation](https://github.com/JuliaParallel/Dagger.jl/blob/e16ea92e358b53a1a5a448dbdaf1c7c8292da9df/src/array/map-reduce.jl#L58)

jariji commented:
Fwiw, [`Folds.reduce`](https://juliafolds.github.io/Folds.jl/dev/#Folds.reduce) documents it assumption that

> binary function `op` is (at least approximately) associative and `init` behaves as the identity of `op`.

mbauman commented:
OK, I would still like to improve our semantics here.  I've opened my shot at take-2 in #53945.

StefanKarpinski commented:
I'm a bit late to the party, but after consideration and discussion with @mbauman, I'm on board with this change. I think the guarantee can be expressed more clearly this way:

> When no `init` value is provided, some `op` calls will take collection values as both arguments. When `init` is provided, values from the collection will only be used as right arguments to `op`, never as left arguments—the left argument of `op` will always be `init` or the result of a previous call to `op`.

I agree that the wording of the new docs is a bit verbose, but we can always worry about that after the functionality is nailed down.
